Server startup error in Windows

**Description:**
Starting up the SP server in any runtime is throwing up the following error. The error appears to be caused by the kernel not resolving the ${sys:carbon.home} property (i.e. missing \ characters):
Bad pathname"; "E:wso2sp-4.0.0-Alpha3-SNAPSHOTbin../wso2/editor/database/WSO2_CARBON_DB.mv.db"

The full error is as below:
[2017-11-24 04:08:52,874] ERROR {org.wso2.carbon.kernel.internal.startupresolver.StartupComponentManager} - RuntimeException occurred while calling onAllRequiredCapabilitiesAvailable of component carbon-datasource-service com.zaxxer.hikari.pool.PoolInitializationException: Exception during pool initialization: IOException: "java.io.IOException: Bad pathname"; "E:wso2sp-4.0.0-Alpha3-SNAPSHOTbin../wso2/editor/database/WSO2_CARBON_DB.mv.db" [90031-196] at com.zaxxer.hikari.pool.HikariPool.initializeConnections(HikariPool.java:581) at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:152) at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:73) at org.wso2.carbon.datasource.rdbms.hikari.HikariRDBMSDataSource.getDataSource(HikariRDBMSDataSource.java:56) at org.wso2.carbon.datasource.rdbms.hikari.HikariDataSourceReader.createDataSource(HikariDataSourceReader.java:74) at org.wso2.carbon.datasource.core.DataSourceBuilder.buildDataSourceObject(DataSourceBuilder.java:79) at org.wso2.carbon.datasource.core.DataSourceBuilder.buildDataSourceObject(DataSourceBuilder.java:60) at org.wso2.carbon.datasource.core.DataSourceBuilder.buildCarbonDataSource(DataSourceBuilder.java:44) at org.wso2.carbon.datasource.core.DataSourceManager.initDataSources(DataSourceManager.java:153) at org.wso2.carbon.datasource.core.internal.DataSourceListenerComponent.onAllRequiredCapabilitiesAvailable(DataSourceListenerComponent.java:125) at org.wso2.carbon.kernel.internal.startupresolver.StartupComponentManager.lambda$notifySatisfiableComponents$7(StartupComponentManager.java:250) at java.util.ArrayList.forEach(ArrayList.java:1249) at org.wso2.carbon.kernel.internal.startupresolver.StartupComponentManager.notifySatisfiableComponents(StartupComponentManager.java:236) at org.wso2.carbon.kernel.internal.startupresolver.StartupOrderResolver$1.run(StartupOrderResolver.java:204) at java.util.TimerThread.mainLoop(Timer.java:555) at java.util.TimerThread.run(Timer.java:505) Caused by: org.h2.jdbc.JdbcSQLException: IO Exception: "java.io.IOException: Bad pathname"; "E:wso2sp-4.0.0-Alpha3-SNAPSHOTbin../wso2/editor/database/WSO2_CARBON_DB.mv.db" [90031-196] at org.h2.message.DbException.getJdbcSQLException(DbException.java:345) at org.h2.message.DbException.get(DbException.java:168) at org.h2.message.DbException.convertIOException(DbException.java:330) at org.h2.store.fs.FilePathDisk.toRealPath(FilePathDisk.java:204) at org.h2.store.fs.FilePathDisk.toRealPath(FilePathDisk.java:34) at org.h2.store.fs.FileUtils.toRealPath(FileUtils.java:77) at org.h2.engine.ConnectionInfo.getName(ConnectionInfo.java:412) at org.h2.engine.Engine.openSession(Engine.java:49) at org.h2.engine.Engine.openSession(Engine.java:176) at org.h2.engine.Engine.createSessionAndValidate(Engine.java:154) at org.h2.engine.Engine.createSession(Engine.java:137) at org.h2.engine.Engine.createSession(Engine.java:27) at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:354) at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:116) at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:100) at org.h2.Driver.connect(Driver.java:69) at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:95) at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:101) at com.zaxxer.hikari.pool.HikariPool.addConnection(HikariPool.java:496) at com.zaxxer.hikari.pool.HikariPool.initializeConnections(HikariPool.java:565) ... 15 more Caused by: java.io.IOException: Bad pathname at java.io.WinNTFileSystem.canonicalize0(Native Method) at java.io.WinNTFileSystem.canonicalize(WinNTFileSystem.java:428) at java.io.File.getCanonicalPath(File.java:618) at org.h2.store.fs.FilePathDisk.toRealPath(FilePathDisk.java:201) ... 31 more

**Suggested Labels:**
Blocker

**Suggested Assignees:**
<!--Optional comma separated list of suggested team members who should attend the issue. Non committers can’t assign issues to assignees, so this will help issue creators who are not a committer to suggest possible assignees-->

**Affected Product Version:**

**OS, DB, other environment details and versions:**    
Windows 7, Windows 10.

**Steps to reproduce:**
Run the server in windows environment

**Related Issues:**
wso2/product-sp#163, wso2/carbon-kernel#1615

--- a/components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/ConfigurationUtils.java
+++ b/components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/ConfigurationUtils.java
@@ -38,6 +38,8 @@ public class ConfigurationUtils {
 
     private static final Logger logger = LoggerFactory.getLogger(ConfigurationUtils.class);
     private static final Pattern varPattern = Pattern.compile("\\$\\{([^}]*)}");
+    private static final char[] specialCharArray = new char[]{'\\', '+', '-', '!', '(', ')', ':', '^', '[', ']',
+            '\"', '{', '}', '~', '*', '?', '|', '&', ';', '/'};
 
     private ConfigurationUtils() {
     }
@@ -162,6 +164,7 @@ public class ConfigurationUtils {
 
     /**
      * Returns replace value with escaped characters.
+     *
      * @param value replaced values
      * @return value with escaped characters
      */
@@ -172,10 +175,14 @@ public class ConfigurationUtils {
         StringBuilder sb = new StringBuilder();
         for (int i = 0; i < value.length(); i++) {
             char c = value.charAt(i);
-            if (c == '\\' || c == '+' || c == '-' || c == '!'  || c == '(' || c == ')' || c == ':'
-                    || c == '^' || c == '[' || c == ']' || c == '\"' || c == '{' || c == '}' || c == '~'
-                    || c == '*' || c == '?' || c == '|' || c == '&'  || c == ';' || c == '/'
-                    || Character.isWhitespace(c)) {
+
+            boolean charExists = false;
+            for (char s : specialCharArray) {
+                if (c == s) {
+                    charExists = true;
+                }
+            }
+            if (charExists || Character.isWhitespace(c)) {
                 sb.append('\\');
             }
             sb.append(c);

--- a/components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/ConfigurationUtils.java
+++ b/components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/ConfigurationUtils.java
@@ -159,4 +159,24 @@ public class ConfigurationUtils {
         }
         return value;
     }
+
+    /**
+     * Returns replace value with escaped characters.
+     * @param value replaced values
+     * @return value with escaped characters
+     */
+    public static String escapeSpecialCharacters(String value) {
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < value.length(); i++) {
+            char c = value.charAt(i);
+            if (c == '\\' || c == '+' || c == '-' || c == '!'  || c == '(' || c == ')' || c == ':'
+                    || c == '^' || c == '[' || c == ']' || c == '\"' || c == '{' || c == '}' || c == '~'
+                    || c == '*' || c == '?' || c == '|' || c == '&'  || c == ';' || c == '/'
+                    || Character.isWhitespace(c)) {
+                sb.append('\\');
+            }
+            sb.append(c);
+        }
+        return sb.toString();
+    }
 }

--- a/components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/ConfigurationUtils.java
+++ b/components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/ConfigurationUtils.java
@@ -166,6 +166,9 @@ public class ConfigurationUtils {
      * @return value with escaped characters
      */
     public static String escapeSpecialCharacters(String value) {
+        if (value == null) {
+            return null;
+        }
         StringBuilder sb = new StringBuilder();
         for (int i = 0; i < value.length(); i++) {
             char c = value.charAt(i);

--- a/components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/provider/ConfigProviderImpl.java
+++ b/components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/provider/ConfigProviderImpl.java
@@ -674,12 +674,14 @@ public class ConfigProviderImpl implements ConfigProvider {
         String newValue = func.apply(key);
         //If the new value is not null, replace the placeholder with the new value and return the string.
         if (newValue != null) {
-            return inputString.replaceFirst(PLACEHOLDER_REGEX, "$1" + newValue + "$8");
+            return inputString.replaceFirst(PLACEHOLDER_REGEX, "$1" +
+                    ConfigurationUtils.escapeSpecialCharacters(newValue) + "$8");
         }
         //If the new value is empty and the default value is not empty, replace the placeholder with the default
         // value and return the string
         if (defaultValue != null) {
-            return inputString.replaceFirst(PLACEHOLDER_REGEX, "$1" + defaultValue + "$8");
+            return inputString.replaceFirst(PLACEHOLDER_REGEX, "$1" +
+                    ConfigurationUtils.escapeSpecialCharacters(defaultValue) + "$8");
         }
         //Otherwise print an error message and throw na exception
         String msg;

--- a/components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/provider/ConfigProviderImpl.java
+++ b/components/org.wso2.carbon.config/src/main/java/org/wso2/carbon/config/provider/ConfigProviderImpl.java
@@ -645,7 +645,8 @@ public class ConfigProviderImpl implements ConfigProvider {
                         SecureVault secureVault = getSecureVault().orElseThrow(() ->
                                 new ConfigurationRuntimeException("Secure Vault service is not available"));
                         String newValue = new String(secureVault.resolve(value));
-                        inputString = inputString.replaceFirst(PLACEHOLDER_REGEX, "$1" + newValue + "$8");
+                        inputString = inputString.replaceFirst(PLACEHOLDER_REGEX, "$1" + ConfigurationUtils
+                                .escapeSpecialCharacters(newValue) + "$8");
                     } catch (SecureVaultException e) {
                         throw new ConfigurationRuntimeException("Unable to resolve the given alias", e);
                     }

--- a/components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/UtilsTest.java
+++ b/components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/UtilsTest.java
@@ -18,6 +18,7 @@ package org.wso2.carbon.config;
 import org.testng.Assert;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
+import org.wso2.carbon.config.annotation.Configuration;
 import org.wso2.carbon.config.utils.EnvironmentUtils;
 
 /**
@@ -111,5 +112,13 @@ public class UtilsTest {
         String config = "${" + UTIL_ENV + "}" + pathSeparator + "deployment" + pathSeparator;
         Assert.assertEquals(ConfigurationUtils.substituteVariables(config),
                 path + pathSeparator + "deployment" + pathSeparator);
+    }
+
+    @Test
+    public void testEscapeSpecialCharacters() {
+        String inputString = "C:\\Software\\WSO2*Kernal-5.0.0\\wso2\\worker\\bin";
+        String expectedString = "C\\:\\\\Software\\\\WSO2\\*Kernal\\-5.0.0\\\\wso2\\\\worker\\\\bin";
+        String outputString = ConfigurationUtils.escapeSpecialCharacters(inputString);
+        Assert.assertEquals(expectedString, outputString);
     }
 }

--- a/components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/UtilsTest.java
+++ b/components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/UtilsTest.java
@@ -120,4 +120,11 @@ public class UtilsTest {
         String outputString = ConfigurationUtils.escapeSpecialCharacters(inputString);
         Assert.assertEquals(expectedString, outputString);
     }
+
+    @Test
+    public void testEscapeSpecialCharactersWithNullValue() {
+        String inputString = null;
+        String outputString = ConfigurationUtils.escapeSpecialCharacters(inputString);
+        Assert.assertNull(outputString);
+    }
 }

--- a/components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/UtilsTest.java
+++ b/components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/UtilsTest.java
@@ -18,7 +18,6 @@ package org.wso2.carbon.config;
 import org.testng.Assert;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
-import org.wso2.carbon.config.annotation.Configuration;
 import org.wso2.carbon.config.utils.EnvironmentUtils;
 
 /**

--- a/components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/UtilsTest.java
+++ b/components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/UtilsTest.java
@@ -113,18 +113,16 @@ public class UtilsTest {
                 path + pathSeparator + "deployment" + pathSeparator);
     }
 
-    @Test
-    public void testEscapeSpecialCharacters() {
-        String inputString = "C:\\Software\\WSO2*Kernal-5.0.0\\wso2\\worker\\bin";
-        String expectedString = "C\\:\\\\Software\\\\WSO2\\*Kernal\\-5.0.0\\\\wso2\\\\worker\\\\bin";
-        String outputString = ConfigurationUtils.escapeSpecialCharacters(inputString);
-        Assert.assertEquals(expectedString, outputString);
+    @DataProvider(name = "params")
+    public Object[][] createParamValues() {
+        return new Object[][]{{"C:\\Software\\WSO2*Kernal-5.0.0\\wso2\\worker\\bin",
+                "C\\:\\\\Software\\\\WSO2\\*Kernal\\-5.0.0\\\\wso2\\\\worker\\\\bin"},
+                {null, null}};
     }
 
-    @Test
-    public void testEscapeSpecialCharactersWithNullValue() {
-        String inputString = null;
+    @Test(dataProvider = "params")
+    public void testEscapeSpecialCharacters(String inputString, String expectedString) {
         String outputString = ConfigurationUtils.escapeSpecialCharacters(inputString);
-        Assert.assertNull(outputString);
+        Assert.assertEquals(expectedString, outputString);
     }
 }

--- a/components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/configprovider/TestUtils.java
+++ b/components/org.wso2.carbon.config/src/test/java/org/wso2/carbon/config/configprovider/TestUtils.java
@@ -17,7 +17,6 @@
  */
 package org.wso2.carbon.config.configprovider;
 
-import org.wso2.carbon.secvault.SecureVaultUtils;
 
 import java.net.URL;
 import java.nio.file.Path;
@@ -40,7 +39,7 @@ public class TestUtils {
      * @return path of the resources
      */
     public static Optional<Path> getResourcePath(String... resourcePaths) {
-        URL resourceURL = SecureVaultUtils.class.getClassLoader().getResource("");
+        URL resourceURL = TestUtils.class.getClassLoader().getResource("");
         if (resourceURL != null) {
             String resourcePath = resourceURL.getPath();
             if (resourcePath != null) {

